code/game/machinery/doors/door.dm

```diff
@@ -67,8 +80,45 @@
 	health = maxhealth
 
 	update_nearby_tiles(need_rebuild=1)
+	if (hashatch)
+		setup_hatch()
 	return
 
+/obj/machinery/door/proc/setup_hatch()
+
+	if (overlays != null)
+		hatch_image = image('icons/obj/doors/hatches.dmi', src, hatchstyle, closed_layer+0.1)
+		hatch_image.color = hatch_colour
+		hatch_image.pixel_x = hatch_offset_x
+		hatch_image.pixel_y = hatch_offset_y
+
+		overlays += hatch_image
+		update_icon()
+	else
+		spawn(10)
+			setup_hatch()
+
+/obj/machinery/door/proc/open_hatch(var/atom/mover = null)
+	if (!hatchstate)
+		hatchstate = 1
+		update_icon()
+		playsound(src.loc, hatch_open_sound, 40, 1, -1)
+	hatchclosetime = world.time + 29
+
+	if (istype(mover, /mob/living/silicon))
+		var/mob/living/silicon/S = mover
+		S.under_door()
+	spawn(30)

```

Comment:
The trouble comes from the fact that the rate at which `world.time` ticks is easily influenced by server load. This is not something that is noticeable on your own machine. Two separate methods of timing can lead to desync, or other various things, if the server is suffering from more strain than normal (or even from regular load). It is, in my mind, a bad (and silly) practice.

If `spawn(30)` works fine, then the if check right after is probably not even necessary.